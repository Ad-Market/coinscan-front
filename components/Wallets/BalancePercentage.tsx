import { utils, BigNumber } from 'ethers';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import LinearProgress, {
  LinearProgressProps,
} from '@mui/material/LinearProgress';
import settings from '../../settings.json';

const balancePercentage = (value: string) => {
  // make balance larger as BigNumber doesn't support fractions
  const balance = BigNumber.from(value).mul(BigNumber.from(1000000));
  const totalSupply = utils.parseUnits(
    settings.totalSupply,
    settings.decimalPlaces
  );
  const percent = balance.div(totalSupply);
  // 4-digit precision
  return percent.toNumber() / 10000;
};

function LinearProgressWithLabel(
  props: LinearProgressProps & { value: number }
) {
  return (
    <Box sx={{ display: 'flex', alignItems: 'center' }}>
      <Box sx={{ width: '100%', mr: 1 }}>
        <LinearProgress variant="determinate" {...props} />
      </Box>
      <Box sx={{ minWidth: 35 }}>
        <Typography
          variant="body2"
          color="text.secondary"
        >{`${props.value}%`}</Typography>
      </Box>
    </Box>
  );
}

const BalancePercentage = ({ balance }: { balance: string }) => {
  return (
    <LinearProgressWithLabel
      variant="determinate"
      value={balancePercentage(balance)}
    />
  );
};

export default BalancePercentage;
