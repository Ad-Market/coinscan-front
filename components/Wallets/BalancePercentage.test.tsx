import { render, screen } from '@testing-library/react';

import BalancePercentage from './BalancePercentage';

test('loads and displays no fractional percent', async () => {
  const { container } = render(
    <BalancePercentage balance={'30000000000000000'} />
  );
  expect(screen.getByText('30%'));
  expect(container).toMatchSnapshot();
});

test('loads and displays fractional percent', async () => {
  const { container } = render(
    <BalancePercentage balance={'1111100000000000'} />
  );
  expect(screen.getByText('1.1111%'));
  expect(container).toMatchSnapshot();
});

test('loads and displays very small percent as 0%', async () => {
  const { container } = render(<BalancePercentage balance={'1000000000'} />);
  expect(screen.getByText('0%'));
  expect(container).toMatchSnapshot();
});
