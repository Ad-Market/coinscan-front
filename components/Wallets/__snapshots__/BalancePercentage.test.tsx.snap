// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads and displays fractional percent 1`] = `
<div>
  <div
    class="MuiBox-root css-70qvj9"
  >
    <div
      class="MuiBox-root css-9vy3tp"
    >
      <span
        aria-label="Percent of supply"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="1"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
      >
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style="transform: translateX(-98.8889%);"
        />
      </span>
    </div>
    <div
      class="MuiBox-root css-pazuet"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        1.1111%
      </p>
    </div>
  </div>
</div>
`;

exports[`loads and displays no fractional percent 1`] = `
<div>
  <div
    class="MuiBox-root css-70qvj9"
  >
    <div
      class="MuiBox-root css-9vy3tp"
    >
      <span
        aria-label="Percent of supply"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="30"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
      >
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style="transform: translateX(-70%);"
        />
      </span>
    </div>
    <div
      class="MuiBox-root css-pazuet"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        30%
      </p>
    </div>
  </div>
</div>
`;

exports[`loads and displays very small percent as 0% 1`] = `
<div>
  <div
    class="MuiBox-root css-70qvj9"
  >
    <div
      class="MuiBox-root css-9vy3tp"
    >
      <span
        aria-label="Percent of supply"
        aria-valuemax="100"
        aria-valuemin="0"
        aria-valuenow="0"
        class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
      >
        <span
          class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style="transform: translateX(-100%);"
        />
      </span>
    </div>
    <div
      class="MuiBox-root css-pazuet"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
      >
        0%
      </p>
    </div>
  </div>
</div>
`;
